controllers/state_of_the_world.go

```diff
@@ -0,0 +1,170 @@
+		ReconcileFunc: istioGatewayProvider.Sample,
+		Events:        append(commonAuthPolicyResourceEventMatchers, controller.ResourceEventMatcher{Kind: ptr.To(IstioAuthorizationPolicyKind)}),
+	}).Reconcile)
+	effectivePolicesReconciler.ReconcileFuncs = append(effectivePolicesReconciler.ReconcileFuncs, (&controller.Subscription{
+		ReconcileFunc: istioGatewayProvider.Sample,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: ptr.To(machinery.GatewayGroupKind), EventType: ptr.To(controller.DeleteEvent)},
+		},
+	}).Reconcile)
+
+	reconciler := &controller.Workflow{
+		Precondition: func(ctx context.Context, resourceEvents []controller.ResourceEvent, _ *machinery.Topology) {
+			logger := controller.LoggerFromContext(ctx).WithName("event logger")
+			for _, event := range resourceEvents {
+				// log the event
+				obj := event.OldObject
+				if obj == nil {
+					obj = event.NewObject
+				}
+				values := []any{
+					"type", event.EventType.String(),
+					"kind", obj.GetObjectKind().GroupVersionKind().Kind,
+					"namespace", obj.GetNamespace(),
+					"name", obj.GetName(),
+				}
+				if event.EventType == controller.UpdateEvent && logger.V(1).Enabled() {
+					values = append(values, "diff", cmp.Diff(event.OldObject, event.NewObject))
+				}
+				logger.Info("new event", values...)
+			}
+		},
+		Tasks: []controller.ReconcileFunc{
+			topologyFileReconciler.Reconcile,
+			effectivePolicesReconciler.Reconcile,
+		},
+	}
+
+	return reconciler.Run
+}
+
+type EffectivePolicesReconciler struct {
+	Client         *dynamic.DynamicClient
+	ReconcileFuncs []controller.ReconcileFunc
+}
+
+func (r *EffectivePolicesReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, _ *machinery.Topology) {
+	logger := controller.LoggerFromContext(ctx).WithName("effective polices reconciler")
+	logger.Info("reconciling effective polices reconciler")
+}
+
+type TopologyFileReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology) {
+	logger := controller.LoggerFromContext(ctx).WithName("topology file")
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "topology",
+			Namespace: "kuadrant-system",
+		},
+		Data: map[string]string{
+			"topology": topology.ToDot(),
+		},
+	}
+
+	configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}
+	unstructuredMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(cm)
+	if err != nil {
+		panic(err.Error())
+	}
+	unstructuredCM := &unstructured.Unstructured{
+		Object: unstructuredMap,
+	}
+
+	_, err = r.Client.Resource(configMapRes).Namespace(cm.Namespace).Update(context.TODO(), unstructuredCM, metav1.UpdateOptions{})
+	if errors.IsNotFound(err) {
+		_, err = r.Client.Resource(configMapRes).Namespace(cm.Namespace).Create(context.TODO(), unstructuredCM, metav1.CreateOptions{})
```

Comment:
I see here a bit of the pattern we want to avoid with the topology, I guess. Rather, I think the controller could be watching ConfigMaps as well, making sure the `kuadrant-system/topology` ConfigMap object is represented in the topology struct (linked to the Kuadrant CR?) Then, calling `Create` vs `Update` would depend on the presence of the object in the topology. WDYT?

Reply:
Okay I have done a bit of a refactor here. I think now it lines up better with what you were expecting.